### DISCUSSION TITLE
Limiting number of tags displayed in a postslist

### DISCUSSION THREAD

**Discussion by sagewaters** [11ty/eleventy#2443] [Q&A] [answered]
Curious how one might limit the number of tags displayed on a Collection page by editing the **postslist include** in the default **11ty starter template.**

Looking at **postslist.njk** file in **_includes**

````
<ol reversed class="postlist" style="counter-reset: start-from {{ (postslistCounter or postslist.length) + 1 }}">
{% for post in postslist | reverse %}
  <li class="postlist-item{% if post.url == url %} postlist-item-active{% endif %}">
    <a href="{{ post.url | url }}" class="postlist-link">{% if post.data.title %}{{ post.data.title }}{% else %}<code>{{ post.url }}</code>{% endif %}</a>
   <time class="postlist-date" datetime="{{ post.date | htmlDateString }}">{{ post.date | readableDate }}</time>
    {% for tag in post.data.tags | filterTagList %}
      {% set tagUrl %}/tags/{{ tag | slug }}/{% endset %}
      <a href="{{ tagUrl | url }}" class="post-tag">{{ tag }}</a>
    {% endfor %}
  </li>
{% endfor %}
</ol>
````

My too-soon-to-call-myself-a-novice approach was to simply add a **"limit:2"**, like so:

```
  {% for tag in post.data.tags limit:2 | filterTagList %}
      {% set tagUrl %}/tags/{{ tag | slug }}/{% endset %}
      <a href="{{ tagUrl | url }}" class="post-tag">{{ tag }}</a>
    {% endfor %}
```

...but that sparked another of errors when 11ty attempted to serve the site up.

I imagine there is a simple solution that I am missing here. Any help would be greatly appreciated.

**Comment by pdehaan** [comment, marked as answer]
Should be fairly easy to write a custom "limit" filter which just pulls the top _n_ items from the head of an array.
Google suggested https://epsi-rns.gitlab.io/ssg/2020/01/19/11ty-widget-nunjucks-loop/

Re: Nunjucks, I didn't see any `limit` mentioned in the `for` loop docs at https://mozilla.github.io/nunjucks/templating.html#for.
I know LiquidJS supports that syntax though: https://liquidjs.com/tags/for.html#Parameters

**Reply by pdehaan** [reply]
In fact, I think [eleventy-base-blog's **.eleventy.js**](https://github.com/11ty/eleventy-base-blog/blob/2d3774c22513c03ea6c53af33f1aec1840b30b9b/.eleventy.js#L30-L40) has this function which is really nice (and lets you get the first _n_ items or last _n_ items from the list):

```js
  // Get the first `n` elements of a collection.
  eleventyConfig.addFilter("head", (array, n) => {
    if(!Array.isArray(array) || array.length === 0) {
      return [];
    }
    if( n < 0 ) {
      return array.slice(n);
    }

    return array.slice(0, n);
  });
```

And you can check the https://github.com/11ty/eleventy-base-blog/blob/2d3774c22513c03ea6c53af33f1aec1840b30b9b/index.njk#L10-L12 for usage:
```njk
{% set postslist = collections.posts | head(-3) %}
{% set postslistCounter = collections.posts | length %}
```

**Reply by sagewaters** [reply]
Thank you, @pdehaan ! Very much appreciated. Giving it a go now.

**Reply by Ryuno-Ki** [reply]
It's called [slice](https://mozilla.github.io/nunjucks/templating.html#slice) in Nunjucks.

**Reply by pdehaan** [reply]
@Ryuno-Ki That always confuses me. I always think `slice` filter would work like JavaScript array `.slice()`, but it returns an ~expected~ _unexpected_ list of lists.

> _"Slice an iterator and return a list of lists containing those items:"_

So if I don't care about partitioned lists and just want the first _n_ items, I have to do weird mixes of `| slice(3) | first`:

```njk
{% set arr = ["one", "two", "three", "four", "five", "six", "seven"] %}

<div class="columwrapper">
  {%- for item in arr | slice(3) | first %}
    <p>{{ item }}</p>
  {%- endfor %}
</div>
```

### OUTPUT
```html
<div class="columwrapper">
    <p>one</p>
    <p>two</p>
    <p>three</p>
</div>
```

---

versus:

```njk
{% set arr = ["one", "two", "three", "four", "five", "six", "seven"] %}

<div class="columwrapper">
  {%- for items in arr | slice(3) %}
    <p>{{ items }}</p>
  {%- endfor %}
</div>
```

### OUTPUT
```html
<div class="columwrapper">
    <p>one,two,three</p>
    <p>four,five</p>
    <p>six,seven</p>
</div>
```

**Reply by Ryuno-Ki** [reply]
Well, it's more flexible :)

**Reply by sagewaters** [reply]
@pdehaan It was indeed staring me in the face as a made-to-order option they already baked into the [eleventy-base-blog's .eleventy.js](https://github.com/11ty/eleventy-base-blog/blob/2d3774c22513c03ea6c53af33f1aec1840b30b9b/.eleventy.js#L30-L40) file. And it works like a charm. 

```
    {% for tag in post.data.tags | filterTagList | head(2) %}
      {% set tagUrl %}/tags/{{ tag | slug }}/{% endset %}
      <a href="{{ tagUrl | url }}" class="post-tag">{{ tag }}</a>

```


Thank you **very** much for assist.
